File: Problemset/cses/Dynamic_Programming/c++/Minimizing_Coins.cpp
#include <bits/stdc++.h>
using namespace std;

const int INF = 1e9;
int n, x;
vector<int> coins;
unordered_map<int, int> memo;

int min_coins(int amount) {
    if (amount == 0) return 0;
    if (amount < 0) return INF;
    if (memo.count(amount)) return memo[amount];

    int ans = INF;
    for (int c : coins) {
        int res = min_coins(amount - c);
        if (res != INF) {
            ans = min(ans, res + 1);
        }
    }

    memo[amount] = ans;
    return ans;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> n >> x;
    coins.resize(n);
    for (int i = 0; i < n; i++) cin >> coins[i];

    int result = min_coins(x);
    if (result == INF) cout << -1 << "\n";
    else cout << result << "\n";

    return 0;
}

File: Problemset/cses/Dynamic_Programming/Python/Minimizing_Coins.py
n, x = map(int, input().split())
coins = list(map(int, input().split()))

INF = 10**9
dp = [INF] * (x + 1)
dp[0] = 0

for amount in range(1, x + 1):
    for c in coins:
        if amount - c >= 0:
            dp[amount] = min(dp[amount], dp[amount - c] + 1)

print(dp[x] if dp[x] != INF else -1)
